### PR TITLE
New feature #15246: Allow fixed function (only PHP) in expression manager

### PR DESCRIPTION
**Pull request by Shnoulle**
New feature #13175: Ability to show calculated values based on all users

- set null to js function in ValidFunction : set this function as static
- static function mean done in PHP return a static part for EM js
- statFunctions plugin with statCount and statCountIf
- related issue [#11589 quota accessible by EM](https://bugs.limesurvey.org/view.php?id=11589)
- related issue [#12248 expose mt_rand](https://bugs.limesurvey.org/view.php?id=12248) (made it more simple)
- test must be done (lsa is here)

### CODE REVIEW

**Comment by Shnoulle** [comment]
I check travis difference when i found time :)

**Comment by Shnoulle** [comment]
OK, travis compare
`Tests: 134, Assertions: 15328, Errors: 1, Failures: 12, Skipped: 6.`
VS
`Tests: 134, Assertions: 15328, Errors: 1, Failures: 12, Skipped: 6.`

I add the plugin test :)

File: application/helpers/expressions/em_core_helper.php [hunk 1]
```diff
@@ -1190,70 +1190,110 @@ public function GetJavaScriptEquivalentOfExpression()
...
+                    case 'WORD':
+                        if ($i + 1 < $numTokens && $tokens[$i + 1][2] == 'LP') {
+                            // then word is a function name
+                            $funcInfo = $this->RDP_ValidFunctions[$token[0]];
+                            if ($funcInfo[1] === null ) {
+                                /* start a static function */
+                                $staticStringToParse = $token[0]; // The function name
+                                $bracket = 0; // Reset bracket (again)
+                            } else {
+                                if ($funcInfo[1] == 'NA') {
```
Comment (Shnoulle) [inline application/helpers/expressions/em_core_helper.php]: Usage of NA for JS function `myfunction("something")` => `"something"`

Maybe remove this checking ?

File: plugins/statFunctions/countFunctions.php [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+     * @return null|string null mean an invalid column/code
+     */
+    private static function _getColumnByQCode($surveyId,$qCode)
+    {
+        $availableColumns = SurveyDynamic::model($surveyId)->getAttributes();
+        /* Sample : Q01.sgqa Q01_SQ01.sgqa */
+        if(array_key_exists($qCode,$availableColumns)) {
+            return $qCode;
+        }
+        /* @todo : allow "Q0" and "Q0_SQ0" …
```
Comment (Shnoulle) [inline plugins/statFunctions/countFunctions.php]: Some are really easy (single choice), some just less simple butbstill easy (Q1_SQ001), some more complex (Q01_other ? Q01comment) etc … 

See https://gitlab.com/SondagesPro/coreAndTools/getQuestionInformation/blob/master/helpers/surveyCodeHelper.php#L74 for sample code to get all …

**Comment by Shnoulle** [comment]
Post-it ;)

**Comment by olleharstedt** [comment]
Yep, it's on the wall, together with general workflow for new features.

**Comment by olleharstedt** [comment]
Can this be tested manually too?

**Comment by Shnoulle** [comment]
Yes,
Activate plugin
Import test survey
Add a response,
Add a response
Values inside survey are updated :)

**Comment by olleharstedt** [comment]
Values are updated how? I would like "idiot safe" instructions so anyone can test, also non-developers.

**Comment by Shnoulle** [comment]
… 

1st pass
Capture d’écran du 2019-10-23 11-02-38

2nd pass
Capture d’écran du 2019-10-23 11-03-59

the plugin name is statcount …

**Comment by olleharstedt** [comment]
Great, will test.

**Comment by olleharstedt** [comment]
image

**Comment by olleharstedt** [comment]
image

**Comment by olleharstedt** [comment]
Don't remember if I showed this picture. Improved workflow for how to merge new features (also for in-house development), to give more stability.

**Comment by Shnoulle** [comment]
Since i do it since some month now …

1. Pull request
2. Ask reviewer
3. Wait … …
4. Merge when i'm sure at 100%

I don't see a «Show a manual test procedure» in your picture.

I think we can add a "unique" answer test too : https://www.limesurvey.org/forum/can-i-do-this-with-limesurvey/119491-limit-one-cnpj-like-ein-per-participant-validation-db

Else : travais seems broken but not related to the last push

**Comment by olleharstedt** [comment]
> I don't see a «Show a manual test procedure» in your picture.

No, some things are still missing that haven't been discussed or thought out, like how a specification for a feature should be written. Are scenarios enough, or do we also want UML class diagrams of architecture and design, etc.

**Comment by Shnoulle** [comment]
> specification for a feature

And can be very hard for plugin events. Herer there are 3 related issues, but this don't fix whole. This offer opportunity to fix …

**Comment by olleharstedt** [comment]
The acceptance test can more or less "be" the specification.

File: .gitignore [hunk 3]
```diff
@@ -77,6 +77,7 @@
 !/plugins/expressionQuestionHelp/
 !/plugins/expressionFixedDbVar/
 !/plugins/expressionQuestionForAll/
+!/plugins/statFunctions/
```
Comment (olleharstedt) [inline .gitignore]: Is "stat" short for "static"? Or "statistics"?

**Comment by olleharstedt** [comment]
Would be nice if all new PHP files would be PSR-2...

File: plugins/statFunctions/countFunctions.php [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+use SurveyDynamic;
+use CDbCriteria;
+use Permission;
+
+class countFunctions
+{
+    /**
+     * Return the count of reponse on current Expression Manager survey equal to a specific value
+     * Can use < or > … see https://www.yiiframework.com/doc/api/1.1/CDbCriteria#compare-detail
+     * @todo
```
Comment (olleharstedt) [inline plugins/statFunctions/countFunctions.php]: Empty todo?

File: .gitignore (hunk 3 above)
Comment (Shnoulle) [reply .gitignore]: statistics, can rename

File: plugins/statFunctions/countFunctions.php (hunk 4 above)
Comment (Shnoulle) [reply plugins/statFunctions/countFunctions.php]: Must remove it :)

File: .gitignore (hunk 3 above)
Comment (Shnoulle) [reply .gitignore]: Maybe statisticsCountFunction if rename

File: plugins/statFunctions/countFunctions.php [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+        if($submitted) {
+            $oCriteria->addCondition("submitdate IS NOT NULL");
+        }
+        return intval(SurveyDynamic::model($surveyId)->count($oCriteria));
+    }
+
+    /**
+     * Check the survey
+     * @param $surveyId
+     * @return integer|string|null null man surveyId can be used
```
Comment (olleharstedt) [inline plugins/statFunctions/countFunctions.php]: What does "null man" mean?

File: .gitignore (hunk 3 above)
Comment (olleharstedt) [reply .gitignore]: But answers are being counted? "AnswerCountFunctions"?

**Comment by olleharstedt** [comment]
As I understand `Q00.NAOK > statCount(Q01.sgqa)`, the question should show when NAOK > number of answers for Q01? I enter "4" and Q01 shows, but there seem to be 8 answers already.

**Comment by olleharstedt** [comment]
I'm going on vacation tomorrow, will be back Wednesday next week and continue the review then. Thanks for the PR, looks good otherwise! Great addition!

File: .gitignore (hunk 3 above)
Comment (Shnoulle) [reply .gitignore]: Yes, but can add statFunctions::medium(…) after …

AnswerCountFunctions : seems more related to Count answers , it's Reponse count …

Comment (Shnoulle) [reply .gitignore]: PS : currently \statFunctions\countFunctions::statCount + \statFunctions\countFunctions::statCountIf , but can add

\statFunctions\mediumFunctions::… or … or … or …

**Comment by Shnoulle** [comment]
> As I understand `Q00.NAOK > statCount(Q01.sgqa)`, the question should show when NAOK > number of answers for Q01? I enter "4" and Q01 shows, but there seem to be 8 answers already.

1. Enter 3 : https://github.com/LimeSurvey/LimeSurvey/pull/1320/files#diff-c865143161e1fb5981adf2d2cf7f9623R52 : Q01 must be hidden
2. Enter 10 : https://github.com/LimeSurvey/LimeSurvey/pull/1320/files#diff-c865143161e1fb5981adf2d2cf7f9623R60 : Q01 must be shown

It work on the test , then : i don't understand why it don't work visually ? I make a develop merge this week and check again.

**Comment by Shnoulle** [comment]
@olleharstedt : i do an error … condition is `Q00.NAOK > statCount(Q01.sgqa)` not `Q00.NAOK < statCount(Q01.sgqa)`

But currently : i can save question … i check to create the lsa manually …

**Comment by Shnoulle** [comment]
> As I understand `Q00.NAOK > statCount(Q01.sgqa)`, the question should show when NAOK > number of answers for Q01? I enter "4" and Q01 shows, but there seem to be 8 answers already.

Fixed in https://github.com/LimeSurvey/LimeSurvey/pull/1320/commits/c0235504b05a9f68200aa0b3b531411c85ce297b

File: plugins/statFunctions/countFunctions.php [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+        }
+        return intval(SurveyDynamic::model($surveyId)->count($oCriteria));
+    }
+
+    /**
+     * Check the survey
+     * @param $surveyId
+     * @return integer|string|null : return 0 or string : value to be directly shown, null : final value must be evaluated
+     */
+    private static function _checkSurveyId($surveyId)
```
Comment (Shnoulle) [inline plugins/statFunctions/countFunctions.php]: Remove this private static function

File: plugins/statFunctions/countFunctions.php [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+        }
+    }
+
+    /**
+     * Get the column name from a code.
+     * @param integer $surveyId
+     * @param string $qCode
+     * @return null|string : null mean an invalid column/code, string is the column in reponse database, null if not found
+     */
+    private static function _getColumnByQCode($surveyId, $qCode)
```
Comment (Shnoulle) [inline plugins/statFunctions/countFunctions.php]: Move this function to an helper

File: plugins/statFunctions/questionCodeHelper.php [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+     */
+    public function __construct($surveyId)
+    {
+        $this->surveyId = $surveyId;
+        /* Throw error if surveyid is invalid ? */
+    }
+
+    /**
+     * Check the survey
+     * @param string $qCode question SGQA
```
Comment (Shnoulle) [inline plugins/statFunctions/questionCodeHelper.php]: Allow EM code is on todo, but need really more work :) (and I think a EM code to SGQA must come from Question::model or theme …)

**Comment by Shnoulle** [comment]
Sh… @olleharstedt : you think the last commit can broke (on PHP7.3, ok on php7.2) ?

**Comment by olleharstedt** [comment]
> Sh… @olleharstedt : you think the last commit can broke other test (on PHP7.3, ok on php7.2) ?

T_T We really shouldn't use Travis anymore. So unpredictable! D:

**Comment by Shnoulle** [comment]
@olleharstedt  : can i be sure to have this in next RC if i move the new plugins in core ???

hat can i update for https://manual.limesurvey.org/How_to_contribute_new_features ? In my pint of view : the test (with plugins) show a great needed user system.

**Comment by olleharstedt** [comment]
When are you available on chat? Any chat, preferably IRC team chat.

**Comment by olleharstedt** [comment]
You need to pull upstream to make Travis green, I think.

**Comment by olleharstedt** [comment]
I would like to merge this, but should discuss internally first about policy.

**Comment by olleharstedt** [comment]
Maybe Carsten wants to release 4.0.0 before we start to merge new features.

**Comment by olleharstedt** [comment]
In which case it will merged into 4.1.0 instead.

File: tests/functional/frontend/FixedFunctionExpressionPluginTest.php [hunk 9]
```diff
@@ -0,0 +1,124 @@
+<?php
```
Comment (olleharstedt) [inline tests/functional/frontend/FixedFunctionExpressionPluginTest.php]: Would be nice to move this to functional/acceptance/15246-fixed-em-function, maybe?

**Comment by Shnoulle** [comment]
Out of office until next week.

Its a «can not broke other part» ... feature ... time where we can add new feature are really short ...

**Comment by maziminke** [comment]
This is a really great new feature. I'd love seeing this at Limesurvey 4!

**Comment by olleharstedt** [comment]
> This is a really great new feature. I'd love seeing this at Limesurvey 4!

It will be. Question is only if in 4.0.0 or 4.x. Probably the latter.

**Comment by Shnoulle** [comment]
Manual way to check (and one of real user usage)

1. Activate statFunctions plugin
2. Import survey from test : survey_archive_statCountFunctionsTest.lsa
3. Check question text : statCount must be blue
4. Launch survey, go to second page : last lines is disabled

Capture d’écran du 2019-12-06 17-09-55

**Comment by Shnoulle** [comment]
@olleharstedt : whan adding a test suite in ` tests/functional/acceptance/****/***Test.php` : is not auito added in Test : right ?

Can you give me the line to update please ?

**Comment by olleharstedt** [comment]
I think I added that folder to phpunit.xml before?

**Comment by olleharstedt** [comment]
This line should be enough to run all tests in acceptance/:
https://github.com/LimeSurvey/LimeSurvey/blob/master/phpunit.xml#L22

**Comment by Shnoulle** [comment]
> This line should be enough to run all tests in acceptance/:
> https://github.com/LimeSurvey/LimeSurvey/blob/master/phpunit.xml#L22


Right … sorry … my error : comparing `vue-cli-service test:unit` number of test, not `PHPUnit` test …

170 vs 169 :+1:  for PHPUnit

**Comment by olleharstedt** [comment]
IMO, this should be merged after release.

**Pull request merged**